crates/interpreter/src/valid.rs

```diff
@@ -560,7 +585,8 @@ impl<'a, 'm> Expr<'a, 'm> {
             Nop => (),
             Block(b) => self.push_label(self.blocktype(&b)?, LabelKind::Block)?,
             Loop(b) => {
-                self.push_label(self.blocktype(&b)?, LabelKind::Loop(self.branch_target()))?
+                let type_ = self.blocktype(&b)?;
+                self.push_label(type_, LabelKind::Loop(self.branch_target(type_.results.len())))?
```

Comment:
For loop, I'm not sure whether I should use `type_.results.len()` (as in other cases) or `type_.params.len()` for `val_cnt`. But I suspect the latter is correct for the following reason: in `exec.rs`, [the last `arity` values are kept](https://github.com/zhouwfang/wasefire/blob/de6c7e889865736cb1f11877d5d4e1e9f50729a8/crates/interpreter/src/exec.rs#L1058-L1059) which should be the same as `val_cnt`. By [the definition of `arity`](https://github.com/zhouwfang/wasefire/blob/de6c7e889865736cb1f11877d5d4e1e9f50729a8/crates/interpreter/src/exec.rs#L1040-L1043) in `exec.rs`, the latter is used.

Reply:
Let me know if my understanding about the reason why it is different for loop is correct: For loop, we jump to the beginning instead of the end. When we jump to the beginning, we need to look at the params. The result of the loop will be handled by the outer scope of the loop.

Reply:
Yes